### DISCUSSION TITLE
Custom Script Help

### DISCUSSION THREAD

**Discussion by IamCoderish** [artilleryio/artillery#3196] [Artillery]
Hello,

I am trying to dynamically pass changing variables in my json API payload every time a VU is created.  I was looking at the `BeforeRequest` option and it seems like this will work.  However I am running into issue with my js function.  I am transpiling my TS code back to JS and running the JS script in the YAML file.  I'm using the faker module [https://fakerjs.dev/](url) to generate fake data to pass into the request payload.

Following the guide [https://www.artillery.io/docs/reference/test-script#processor---load-custom-code](url) I am was calling `next()` at the end of my function, but I keep getting error `⠴ worker error, id: 1 TypeError [Error]: next is not a function` which is obvious since next is just param.  I saw in a different discussion [https://github.com/artilleryio/artillery/discussions/2638](url) that they were able to get this to work w/o error.  I know I am really close to getting this to work, any help would be appreciated.

`yamlParser.ts` 
```typescript
{
  const faker = new Faker();
  const variables = {
    birthdate: await faker.generateFormattedBirthdate(18, 70, "age"),
    state: await faker.fakerState(true),
    gender: await faker.fakerGender(),
    faceAmount: await faker.fakerRandomDollarAmount(50000, 10000000),
    healthClass: await faker.randomNumberGenerator(["PP", "P", "S"]),
    term: await faker.randomNumberGenerator([5, 7, 10, 15, 20, 25, 30, 35, 40]),
  };
  context.vars = JSON.stringify(variables);
  return next();
}
``` 
`yaml`
``` environments:
    qa:
      target: https://<site.com>
      phases:
        - duration: 10m
          arrivalRate: 1
          rampTo: 2.5
          name: Warm up phase
        - pause: 5m
          name: 5min Pause
        - duration: 12m
          arrivalRate: 1.5
          rampTo: 3
          name: Ramp up load
        - pause: 3m
          name: 3min Pause
        - duration: 2m
          arrivalRate: 4
          name: Spike phase
      processor: >-
       JS/utilities/Performance/yamlParser.js
scenarios:
  - name: Check Endpoint Availability
    flow:
      - post:
          url: /api/v1/endpoint
          beforeRequest: updateVariables
          json:
            birthdate: { { birthdate } }
            state: { { state } }
            gender: { { gender } }
            faceAmount: { { faceAmount } }
            healthClass: { { healthClass } }
            term: { { term } }
            heightFt: "6"
            heightIn: "0"
            weight: "200"
```

**Comment by hassy** [comment]
@IamCoderish it looks like your custom function is `async` - in that case it does not receive a `next` callback. You can remove the last line of your function and it should all work

**Reply by IamCoderish** [reply]
Hi @hassy thank you for the reply!

I removed the last line and it's running w/o errors now.  However, when I go to make the API call, the call itself is never actually made.  No response code is returned.  This is the output:

`--------------------------------------
Metrics for period to: 07:33:20(-0600) (width: 9.002s)
--------------------------------------

vusers.created: ................................................................ 10
vusers.created_by_name.Check Quote Availability: ........................... 10


Warning: multiple batches of metrics for period 1718371990000 2024-06-14T13:33:10.000Z`

I am wondering if I am still missing something in my function.  This the updated version

```typescript
export async function updateVariables(
  context: any,
  ee: any,
) {
  const faker = new Faker();
  const variables = {
    birthdate: await faker.generateFormattedBirthdate(18, 70, "age"),
    state: await faker.fakerState(true),
    gender: await faker.fakerGender(),
    faceAmount: await faker.fakerRandomDollarAmount(50000, 10000000),
    healthClass: await faker.randomNumberGenerator(["PP", "P", "S"]),
    term: await faker.randomNumberGenerator([5, 7, 10, 15, 20, 25, 30, 35, 40]),
  };
  context.vars = variables;
```

**Reply by IamCoderish** [reply]
Thinking out loud, could this not be working because I am transpiling back to JS instead of just writing in JS and taking Typescript out of the equation?

**Reply by Beliefg** [reply]
@IamCoderish Shouldn't  context.vars = variables; be  instead **context.vars.variables = variables;**
Example:
`function myBeforeRequestHook(req, context, ee, next) {
  context.vars.password = context.funcs.$randPassword();
  next();
}`
